File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    start = min(times)
    end = n * max(times)
    
    answer = end
    
    while start <= end:
        mid = (start + end) // 2
        result = 0
        for time in times:
            result += mid // time
        if result >= n:
            answer = min(answer, mid)
            end = mid - 1
        else:
            start = mid + 1
    
    return answer

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;
class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        long lt = 1, rt = (long) times[times.length - 1] * n;
        long answer = rt;
        while (lt <= rt) {
            long mid = (lt + rt) / 2;
            
            long num = 0;
            
            for (int time: times) {
                num += mid / time;
            }
            
            if (num >= n) {
                answer = mid;
                rt = mid - 1;
            } else {
                lt = mid + 1;
            }
            
        }
        return answer;
    }
    
}